### ISSUE TITLE
[BUG] dpgen is stucked in the stage of "data stating... (this step may take long time)"

### ISSUE THREAD

**Issue by LingjunWu92** [deepmodeling/deepmd-kit#5117]
### Bug summary

Dear Developers,

I found that strangely my dpgen run is stucked at the stage of "data stating... (this step may take long time)", and sometimes lasted for even 20 hours (I don't know what will happen after 20 hours as I just killed the job then), and while it stucked, no outputs were produced, also no errors were reported either. However, the init_bulk stage is totally fine, and I have also tried the examples downloaded from the deepmd website, which turned out absolutely successful either in the init_bulk stage or in the run stage, so it seems very strange why my task just stucked at the run stage. I uploaded my input and ouput data as attached, could you please give me some advice on what might go wrong? Thank you so much.

Best  wishes,
Lingjun

### DeePMD-kit Version

3.1.1

### Backend and its version

TensorFlow 2.19.1

### How did you download the software?

Offline packages

### Input Files, Running Commands, Error Log, etc.

[issue.zip](https://github.com/user-attachments/files/24364260/issue.zip)

The attached zip file includes all related files for the dpgen task.

### Steps to Reproduce

I just used "dpgen init_bulk param.json machine.json" for the init_bulk stage, and "dpgen run param.json machine.json" for the run stage. The param.json and machine.json files for each stage are uploaded before in the zip file.

**Comment by LingjunWu92** [comment]
Dear dosu @dosu, I tried to run with only one valid system in sys_configs, however, it still doesn't work. Could you please tell me what might be wrong and how to enable verbose or debug logging in dpgen? Many thanks.

Best wishes,
Lingjun

**Comment by njzjz** [comment]
I reproduced this on the local RTX 5090 node with the attached `issue3.zip`.

Environment used for the reproducer:

```text
GPU: NVIDIA GeForce RTX 5090
Driver: 580.95.05
Command: srun --gres=gpu:1 dp --tf train input.json
DeePMD-kit: 3.1.1
Backend: TensorFlow 2.19.1
Build variant: cuda
```

The job is not stuck in DP-GEN or in data loading. With the attached `002/input.json`, the run reaches:

```text
Neighbor statistics: training data with minimal neighbor distance: 0.897350
Neighbor statistics: training data with maximum neighbor size: [16  4  4 12  8]
...
data stating... (this step may take long time)
```

Attaching `gdb` to the hung process shows the TensorFlow session is executing `ProdEnvMatA` and the worker thread is inside the GPU neighbor-list builder:

```text
tensorflow::DirectSession::Run(...)
...
ProdEnvMatAOp<Eigen::GpuDevice, double>
_prepare_coord_nlist_gpu<double>(...)
deepmd::build_nlist_gpu<double>(...)
cudaLaunchKernel(...)
```

The CUDA bug appears to be in `source/lib/src/gpu/neighbor_list.cu`, in `parallel_prefix_scan`. The old loop is:

```cpp
for (int ii = threadIdx.x; ii < nall; ii += THREADS_PER_BLOCK) {
  ...
  BlockScan(temp_storage).ExclusiveSum(o_data, o_data, prefix_op);
  __syncthreads();
  ...
}
```

`BlockScan` and `__syncthreads()` are block-wide collectives, but when `nall > THREADS_PER_BLOCK` and the final segment is partial, only threads with `threadIdx.x < nall % THREADS_PER_BLOCK` enter the second iteration. That is divergent use of a block-wide collective and can deadlock.

This input triggers that condition because PBC coordinate copy expands the 12-atom frame to more than one 256-thread segment before building the neighbor list.

I verified this with a standalone CUDA reproducer on the same RTX 5090:

```text
bad/original prefix-scan kernel, nall=324: timeout after 8 s, exit 124
fixed prefix-scan kernel, nall=324: returns immediately
```

The fix is to make every thread enter every segment iteration, and guard only the memory access:

```cpp
for (int base = 0; base < nall; base += THREADS_PER_BLOCK) {
  int ii = base + threadIdx.x;
  int i_data = ii < nall ? temp_nlist[block_offset + ii] : -1;
  ...
  BlockScan(temp_storage).ExclusiveSum(o_data, o_data, prefix_op);
  __syncthreads();
  if (ii < nall && i_data != -1) {
    nei_order[block_offset + ii] = o_data;
  }
  if (ii == nall - 1) {
    ...
  }
}
```

I also added a GPU unit-test case locally for `nall = TPB + 68` so this tail-segment case is covered.

Workarounds remain the same until this is patched in a release: use the PyTorch backend, run the TensorFlow backend on CPU/non-affected path, or avoid the TF CUDA `ProdEnvMatA` path on RTX 5090/5090D.

**Comment by njzjz** [comment]
I verified the fix with a CUDA build from current master plus the patch and opened PR #5575: https://github.com/deepmodeling/deepmd-kit/pull/5575

Validation details:
- Built from source with CUDA 12.9, TensorFlow 2.19.1, and `CMAKE_CUDA_ARCHITECTURES=120`.
- Confirmed the runtime loaded the freshly built native libraries, including `libdeepmd_op_cuda.so`.
- Re-ran the issue input on an RTX 5090 with `srun --gres=gpu:1 dp --tf train input.json`.
- The run passed the previous `data stating...` hang point, reached `built lr`, `built network`, `built training`, and completed `stop_batch = 2000` with exit code 0.

Root cause: `parallel_prefix_scan` used a per-thread loop around block-wide collectives (`cub::BlockScan` and `__syncthreads()`). For a partial tail segment when `nall` is not divisible by the block size, only part of the block entered the final iteration, which can deadlock the GPU op. The PR makes every thread enter every segment iteration and masks out inactive tail lanes.

**Comment by njzjz** [comment]
FYI @denghuilu

**Comment by njzjz** [comment]
One additional note about why this was seen on RTX 5090 but not necessarily on older GPUs:

I do not think this is a 5090-specific correctness issue. The original kernel already had undefined/invalid CUDA synchronization behavior: the tail iteration of `parallel_prefix_scan` could be entered by only part of a thread block, while it still called block-wide collectives (`cub::BlockScan` and `__syncthreads()`). CUDA requires all non-exited threads in the block to participate consistently in these collectives; otherwise the behavior is undefined and may hang.

The 5090 / SM 12.0 environment likely makes this latent bug much easier to trigger or observe. The generated code, CUB implementation details, driver JIT path, and block scheduling/barrier behavior can differ between GPU architectures and CUDA versions. On older GPUs the same invalid pattern may have appeared to work by accident, or may only have been masked by different code generation/timing. On 5090 it reliably manifests as a GPU op stuck in `build_nlist_gpu`.

So the fix should not special-case 5090. The correct fix is to make every thread in the block enter every `BlockScan`/`__syncthreads()` iteration and mask out inactive tail lanes, which is what PR #5575 does.

**Comment by taozhisunsysu-coder** [comment]
Great work on catching this bug! Any rough ETA on when this PR will be merged and included in a new official release?
